server/src/main/java/org/opensearch/gateway/ReplicaShardBatchAllocator.java

```diff
@@ -0,0 +1,200 @@
+                shardAllocationDecisions.put(shard, AllocateUnassignedDecision.NOT_TAKEN);
+                continue;
+            }
+
+            Tuple<Decision, Map<String, NodeAllocationResult>> result = canBeAllocatedToAtLeastOneNode(shard, allocation);
+            Decision allocationDecision = result.v1();
+            if (allocationDecision.type() != Decision.Type.YES && (!explain || !hasInitiatedFetching(shard))) {
+                // only return early if we are not in explain mode, or we are in explain mode but we have not
+                // yet attempted to fetch any shard data
+                logger.trace("{}: ignoring allocation, can't be allocated on any node", shard);
+                shardAllocationDecisions.put(
+                    shard,
+                    AllocateUnassignedDecision.no(
+                        UnassignedInfo.AllocationStatus.fromDecision(allocationDecision.type()),
+                        result.v2() != null ? new ArrayList<>(result.v2().values()) : null
+                    )
+                );
+                continue;
+            }
+            // storing the nodeDecisions in nodeAllocationDecisions if the decision is not YES
+            // so that we don't have to compute the decisions again
+            nodeAllocationDecisions.put(shard, result);
+
+            eligibleShards.add(shard);
+        }
+
+        // Do not call fetchData if there are no eligible shards
+        if (eligibleShards.isEmpty()) {
+            return shardAllocationDecisions;
+        }
+        // only fetch data for eligible shards
+        final FetchResult<NodeStoreFilesMetadataBatch> shardsState = fetchData(eligibleShards, ineligibleShards, allocation);
+
+        for (ShardRouting unassignedShard : eligibleShards) {
+            Tuple<Decision, Map<String, NodeAllocationResult>> result = nodeAllocationDecisions.get(unassignedShard);
+            shardAllocationDecisions.put(
+                unassignedShard,
+                getAllocationDecision(
+                    unassignedShard,
+                    allocation,
+                    convertToNodeStoreFilesMetadataMap(unassignedShard, shardsState),
+                    result,
+                    logger
+                )
+            );
+        }
+        return shardAllocationDecisions;
+    }
+
+    private Map<DiscoveryNode, StoreFilesMetadata> convertToNodeStoreFilesMetadataMap(
+        ShardRouting unassignedShard,
+        FetchResult<NodeStoreFilesMetadataBatch> data
+    ) {
+        if (!data.hasData()) {
+            return null;
+        }
+
+        Map<DiscoveryNode, StoreFilesMetadata> map = new HashMap<>();
+
+        data.getData().forEach((key, value) -> {
+            Map<ShardId, NodeStoreFilesMetadata> batch = value.getNodeStoreFilesMetadataBatch();
+            NodeStoreFilesMetadata metadata = batch.get(unassignedShard.shardId());
+            if (metadata != null && metadata.getStoreFileFetchException() == null) {
```

Comment:
Similarly exception is not handled properly in ```PrimarySHardBatchAllocator``` @amkhar

Reply:
`PrimaryShardAllocator` uses those store error(already existing in non-batch mode) during decision making flow, I didn't want to change that. So, store level exceptions are propagated as it is in primary. 

This store exception is different than Primary response's exception. Primary's response exception is still sent today in non-batch mode flow. This exception in replica-batch mode is what we've added, which was never being used in decision making flow of RSA. So, it needs to be filtered before reaching RSA's `getAllocationDecision`.

Any high level exceptions(except the one which used to be the part of response object) which were being thrown from primary transport will not be sent to PSA's `getAllocationDecision` also, to maintain the same behaviour.

Reply:
For a single shard in RSA, if there was exception during fetch, the request itself would have failed, here node response for a shard (which threw exception) would be swallowed?

Reply:
>if there was exception during fetch, the request itself would have failed

We have 3 types of exceptions coming from transport in case of primary. 
And 2 types of exceptions coming from replica transport. (In case of store failure exceptions, we return [an empty response from replica transport](https://github.com/amkhar/OpenSearch/blob/7a6f8b0eead5371fa18d49534ac76edf199e9c4c/server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataHelper.java#L79-L85)) 


Let's talk about replica here. 

When code itself throws an exception, that result in a failedNode case, i.e. fetch from a particular node has failed. Now, based on the exception type it would be categorized as retry-able or non-retry-able.
1. Retry-able ones - OpenSearchRejectedExecutionException, ReceiveTimeoutTransportException, OpenSearchTimeoutException. We mark that node entry as fetching not done. And will retry in next fetching round.
2. Non retry-able ones - everything else. We store the failure of the exception and mark fetching as done. And this failure [is not being used](https://github.com/amkhar/OpenSearch/blob/7a6f8b0eead5371fa18d49534ac76edf199e9c4c/server/src/main/java/org/opensearch/gateway/AsyncShardFetchCache.java#L147-L156) when cache returns the data to PSA/RSA .


>here node response for a shard (which threw exception) would be swallowed?

ShardBatchCache class which would providing the data to this allocator, would anyway maintain the [same behaviour of not providing exception related responses](https://github.com/opensearch-project/OpenSearch/pull/12504/files#diff-f22c9cb55b850db5c383ce5ebfa022765e7089d220e80bd216b208cc0e7864fcR332-R336) to allocator.  And [retrying the shard](https://github.com/opensearch-project/OpenSearch/pull/12504/files#diff-f22c9cb55b850db5c383ce5ebfa022765e7089d220e80bd216b208cc0e7864fcR238-R248) if there is any retry-able exception thrown from transport.


@shiv0408  - Ideally this null check can be removed.

Reply:
removed the null check

Reply:
Thanks @shiv0408 , looks good now.